File: libphx/src/Window.cpp
#include "PhxMemory.h"
#include "OpenGL.h"
#include "SDLext.h"
#include "PhxString.h"
#include "Vec2.h"
#include "Viewport.h"
#include "Window.h"
#include "WindowMode.h"
#include "Tex2D.h"
#include "Resource.h"

struct Window {
  SDL_Window* handle;
  SDL_GLContext context;
  WindowMode mode;
  SDL_Cursor* cursor;
};

Window* Window_Create (cstr title, int x, int y, int sx, int sy, WindowMode mode) {
  Window* self = MemNew(Window);
  mode |= (SDL_WINDOW_OPENGL | SDL_WINDOW_ALLOW_HIGHDPI);
  cstr titleWithBuild = StrAdd(title, " (libphx version " __DATE__ ")");
  self->handle = SDL_CreateWindow(titleWithBuild, x, y, sx, sy, mode);
  StrFree(titleWithBuild);
  self->context = SDL_GL_CreateContext(self->handle);
  self->mode = mode;
  if (!self->context)
    Fatal("Failed to create OpenGL context for window");
  OpenGL_Init();
  return self;
}

void Window_Free (Window* self) {
  SDL_FreeCursor(self->cursor);   // Can take NULL
  SDL_GL_DeleteContext(self->context);
  SDL_DestroyWindow(self->handle);
  MemFree(self);
}

void Window_BeginDraw (Window* self) {
  Vec2i size;
  SDL_GL_MakeCurrent(self->handle, self->context);
  Window_GetSize(self, &size);
  Viewport_Push(0, 0, size.x, size.y, true);
}

void Window_EndDraw (Window* self) {
  Viewport_Pop();
  SDL_GL_SwapWindow(self->handle);
}

void Window_GetSize (Window* self, Vec2i* out) {
  SDL_GetWindowSize(self->handle, &out->x, &out->y);
}

void Window_GetPosition (Window* self, Vec2i* out) {
  SDL_GetWindowPosition(self->handle, &out->x, &out->y);
}

cstr Window_GetTitle (Window* self) {
  return SDL_GetWindowTitle(self->handle);
}

void Window_SetFullscreen (Window* self, bool fs) {
  SDL_SetWindowFullscreen(self->handle, fs ? WindowMode_Fullscreen : 0);
}

void Window_SetPosition (Window* self, WindowPos x, WindowPos y) {
  SDL_SetWindowPosition(self->handle, x, y);
}

void Window_SetSize (Window* self, int sx, int sy) {
  SDL_SetWindowSize(self->handle, sx, sy);
}

void Window_SetTitle (Window* self, cstr title) {
  SDL_SetWindowTitle(self->handle, title);
}

void Window_SetVsync (Window*, bool vsync) {
  SDL_GL_SetSwapInterval(vsync ? 1 : 0);
}

void Window_SetWindowGrab (Window* self, bool grabbed) {
  SDL_SetWindowGrab(self->handle, grabbed ? 1 : 0);
}

void Window_SetCursor(Window* self, cstr name, int hotx, int hoty) {
  SDL_FreeCursor(self->cursor);   // Can take NULL

  cstr path = Resource_GetPath(ResourceType_Tex2D, name);

  int width, height, components;
  uchar* data = Tex2D_LoadRaw(path, &width, &height, &components);

  SDL_PixelFormatEnum format = (components == 3) ? SDL_PIXELFORMAT_RGB24 : SDL_PIXELFORMAT_RGBA32;
  SDL_Surface* surface = SDL_CreateRGBSurfaceWithFormatFrom(data, width, height, components * 8, width * components, format);
  if (!surface)
    Fatal("Failed to create custom cursor surface for window");

  self->cursor = SDL_CreateColorCursor(surface, hotx, hoty);
  if (!self->cursor)
  {
    SDL_FreeSurface(surface);
    Fatal("Failed to create custom cursor for window");
  }

  SDL_FreeSurface(surface);
  SDL_SetCursor(self->cursor);
}

void Window_ToggleFullscreen (Window* self) {
  if (self->mode & WindowMode_Fullscreen)
    SDL_SetWindowFullscreen(self->handle, 0);
  else
    SDL_SetWindowFullscreen(self->handle, WindowMode_Fullscreen);
  self->mode ^= WindowMode_Fullscreen;
}

void Window_Hide (Window* self) {
  SDL_HideWindow(self->handle);
}

void Window_Show (Window* self) {
  SDL_ShowWindow(self->handle);
}

File: engine/lib/phx/src/window.rs
use crate::common::*;
use crate::internal::*;
use crate::math::IVec2;
use crate::open_gl::*;
use crate::resource::*;
use crate::resource_type::*;
use crate::tex2d_load::*;
use crate::viewport::*;
use crate::window_mode::*;
use crate::window_pos::WindowPos;

use sdl2_sys::*;

#[repr(C)]
pub struct Window {
    pub handle: *mut SDL_Window,
    pub context: SDL_GLContext,
    pub mode: WindowMode,
    pub cursor: *mut SDL_Cursor,
}

impl Drop for Window {
    fn drop(&mut self) {
        unsafe {
            SDL_FreeCursor(self.cursor); // Can take null.
            SDL_GL_DeleteContext(self.context);
            SDL_DestroyWindow(self.handle);
        }
    }
}

#[luajit_ffi_gen::luajit_ffi(managed = true)]
impl Window {
    #[bind(name = "Create")]
    pub fn new(
        title: &str,
        x: WindowPos,
        y: WindowPos,
        sx: i32,
        sy: i32,
        mode: WindowMode,
    ) -> Self {
        unsafe {
            let modeComplete = mode | SDL_WindowFlags::SDL_WINDOW_OPENGL as WindowMode;
            // | SDL_WindowFlags::SDL_WINDOW_ALLOW_HIGHDPI as WindowMode;
            let handle = SDL_CreateWindow(
                title.as_ptr() as *const libc::c_char,
                x,
                y,
                sx,
                sy,
                modeComplete,
            );
            let context = SDL_GL_CreateContext(handle);
            if context.is_null() {
                CFatal!("Failed to create OpenGL context for window");
            }
            OpenGL_Init();

            Self {
                handle,
                context,
                mode: modeComplete,
                cursor: std::ptr::null_mut(),
            }
        }
    }

    pub fn begin_draw(&self) {
        unsafe {
            let mut size: IVec2 = IVec2::ZERO;

            SDL_GL_MakeCurrent(self.handle, self.context);
            self.get_size(&mut size);
            Viewport_Push(0, 0, size.x, size.y, true);
        }
    }

    pub fn end_draw(&self) {
        unsafe {
            Viewport_Pop();
            SDL_GL_SwapWindow(self.handle);
        }
    }

    pub fn get_position(&self, out: &mut IVec2) {
        unsafe {
            SDL_GetWindowPosition(self.handle, &mut out.x, &mut out.y);
        }
    }

    pub fn get_size(&self, out: &mut IVec2) {
        unsafe {
            SDL_GetWindowSize(self.handle, &mut out.x, &mut out.y);
        }
    }

    pub fn get_title(&self) -> String {
        unsafe { SDL_GetWindowTitle(self.handle) }.as_string()
    }

    pub fn set_fullscreen(&self, fs: bool) {
        unsafe {
            SDL_SetWindowFullscreen(self.handle, if fs { WindowMode_Fullscreen } else { 0 });
        }
    }

    pub fn set_position(&self, x: WindowPos, y: WindowPos) {
        unsafe {
            SDL_SetWindowPosition(self.handle, x, y);
        }
    }

    pub fn set_size(&self, sx: i32, sy: i32) {
        unsafe {
            SDL_SetWindowSize(self.handle, sx, sy);
        }
    }

    pub fn set_title(&self, title: &str) {
        unsafe {
            SDL_SetWindowTitle(self.handle, title.as_ptr() as *const libc::c_char);
        }
    }

    pub fn set_vsync(&self, vsync: bool) {
        unsafe {
            SDL_GL_SetSwapInterval(if vsync { 1 } else { 0 });
        }
    }

    pub fn set_cursor(&mut self, name: &str, hotx: i32, hoty: i32) {
        unsafe {
            SDL_FreeCursor(self.cursor); // Can take null.

            let path = Resource_GetPath(ResourceType_Tex2D, name.as_ptr() as *const libc::c_char);

            let mut width: i32 = 0;
            let mut height: i32 = 0;
            let mut components: i32 = 0;
            let data = Tex2D_LoadRaw(path, &mut width, &mut height, &mut components);

            let pixelFormat = if components == 3 {
                SDL_PixelFormatEnum::SDL_PIXELFORMAT_RGB24
            } else {
                SDL_PixelFormatEnum::SDL_PIXELFORMAT_ABGR8888
            };
            let surface = SDL_CreateRGBSurfaceWithFormatFrom(
                data as *mut _,
                width,
                height,
                components * 8,
                width * components,
                pixelFormat as u32,
            );
            if surface.is_null() {
                CFatal!("Failed to create custom cursor surface for window");
            }

            self.cursor = SDL_CreateColorCursor(surface, hotx, hoty);
            if self.cursor.is_null() {
                SDL_FreeSurface(surface);
                CFatal!("Failed to create custom cursor for window");
            }

            SDL_FreeSurface(surface);
            SDL_SetCursor(self.cursor);
        }
    }

    pub fn set_mouse_position(&self, position: &IVec2) {
        unsafe {
            SDL_WarpMouseInWindow(self.handle, position.x, position.y);
        }
    }

    pub fn set_window_grab(&self, grabbed: bool) {
        unsafe {
            SDL_SetWindowGrab(
                self.handle,
                if grabbed {
                    SDL_bool::SDL_TRUE
                } else {
                    SDL_bool::SDL_FALSE
                },
            );
        }
    }

    pub fn toggle_fullscreen(&mut self) {
        unsafe {
            if self.mode & WindowMode_Fullscreen != 0 {
                SDL_SetWindowFullscreen(self.handle, 0);
            } else {
                SDL_SetWindowFullscreen(self.handle, WindowMode_Fullscreen);
            }
            self.mode ^= WindowMode_Fullscreen;
        }
    }

    pub fn hide(&self) {
        unsafe {
            SDL_HideWindow(self.handle);
        }
    }

    pub fn show(&self) {
        unsafe {
            SDL_ShowWindow(self.handle);
        }
    }
}
